File: app/src/main/java/com/mrdoob/browser/PanelRenderer.kt
package com.mrdoob.browser

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Configuration
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import org.json.JSONObject
import org.json.JSONTokener

class PanelRenderer(
    private val activity: MainActivity,
    private val panelView: WebView,
    private val mainView: WebView
) {
    enum class Tab { SOURCE, NETWORK, THREEJS }

    private var shellLoaded = false
    private val pending = ArrayDeque<() -> Unit>()

    @SuppressLint("SetJavaScriptEnabled")
    fun init() {
        panelView.settings.javaScriptEnabled = true
        panelView.settings.mixedContentMode = WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE
        panelView.setBackgroundColor(panelView.materialColor(com.google.android.material.R.attr.colorSurfaceContainer))
        panelView.webChromeClient = WebChromeClient()
        panelView.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView, url: String?) {
                if (shellLoaded) return
                shellLoaded = true
                applyTheme()
                while (pending.isNotEmpty()) pending.removeFirst().invoke()
            }
        }
        val html = buildShellHtml()
        panelView.loadDataWithBaseURL(
            "https://cdn.jsdelivr.net/", html, "text/html", "UTF-8", null
        )
    }

    fun setTab(tab: Tab) {
        afterShellLoaded {
            val name = if (tab == Tab.SOURCE) "source" else "network"
            evalPanel("window.__panel.setActiveTab(${q(name)})")
        }
    }

    fun refreshSource() {
        mainView.evaluateJavascript("document.documentElement.outerHTML") { json ->
            val raw = try {
                JSONTokener(json).nextValue() as? String
            } catch (_: Exception) { null } ?: return@evaluateJavascript
            afterShellLoaded {
                evalPanel("window.__panel.setSource(${q(raw)})")
            }
        }
    }

    fun appendNetworkRow(record: NetworkLog.Record, displayUrl: String) {
        val payload = JSONObject().apply {
            put("method", record.method)
            put("url", record.url)
            put("displayUrl", displayUrl)
            put("status", record.status)
            put("body", record.body)
            put("contentType", record.contentType ?: "")
            put("durationMs", record.durationMs)
            put("size", record.sizeBytes)
            put("requestId", record.requestId ?: "")
        }.toString()
        afterShellLoaded {
            evalPanel("window.__panel.appendNetworkRow($payload)")
        }
    }

    fun clearNetwork() {
        afterShellLoaded {
            evalPanel("window.__panel.clearNetwork()")
        }
    }

    fun deliverBlob(rid: String, dataUrl: String) {
        afterShellLoaded {
            evalPanel("window.__panel.deliverBlob(${q(rid)}, ${q(dataUrl)})")
        }
    }

    private fun applyTheme() {
        val dark = isNightMode()
        val fg = if (dark) "#cccccc" else "#222222"
        val panelBg = panelView.materialColorHex(com.google.android.material.R.attr.colorSurfaceContainer)
        val viewerBg = panelView.materialColorHex(com.google.android.material.R.attr.colorSurfaceContainerHighest)
        evalPanel(
            "window.__panel.setTheme({ dark: $dark, fg: ${q(fg)}, panelBg: ${q(panelBg)}, viewerBg: ${q(viewerBg)} })"
        )
    }

    private fun afterShellLoaded(action: () -> Unit) {
        if (shellLoaded) action() else pending.addLast(action)
    }

    private fun evalPanel(js: String) {
        panelView.evaluateJavascript(js, null)
    }

    private fun q(s: String): String = JSONObject.quote(s)

    private fun isNightMode(): Boolean = (activity.resources.configuration.uiMode and
        Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES

    private fun buildShellHtml(): String {
        val context: Context = activity
        val shell = context.assets.open("panel-shell.html").bufferedReader().use { it.readText() }
        val js = context.assets.open("panel.js").bufferedReader().use { it.readText() }
        return shell.replace("<!--__PANEL_JS__-->", "<script>$js</script>")
    }
}

File: iOS/NetexIOS/Sources/NetexIOS/PanelRenderer.swift
import UIKit
import WebKit

final class PanelRenderer: NSObject, WKNavigationDelegate {
    enum Tab: String {
        case console
        case source
        case network
    }

    private weak var panelView: WKWebView?
    private weak var pageView: WKWebView?
    private var shellLoaded = false
    private var pending: [() -> Void] = []

    init(panelView: WKWebView, pageView: WKWebView) {
        self.panelView = panelView
        self.pageView = pageView
        super.init()
        panelView.navigationDelegate = self
    }

    func load() {
        guard let panelView else { return }
        let shell = AssetLoader.text("panel-shell", ext: "html")
        let script = AssetLoader.text("panel", ext: "js")
            .replacingOccurrences(of: "AndroidPanel.postMessage", with: "window.webkit.messageHandlers.panel.postMessage")
        let html = shell.replacingOccurrences(of: "<!--__PANEL_JS__-->", with: "<script>\(script)</script>")
        panelView.loadHTMLString(html, baseURL: AssetLoader.assetURL("NetexAssets/"))
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        guard !shellLoaded else { return }
        shellLoaded = true
        applyTheme()
        pending.forEach { $0() }
        pending.removeAll()
    }

    func setTab(_ tab: Tab) {
        afterLoad { [weak self] in
            self?.evaluate("window.__panel.setActiveTab('\(tab.rawValue)')")
        }
    }

    func refreshSource() {
        pageView?.evaluateJavaScript("document.documentElement.outerHTML") { [weak self] value, _ in
            guard let html = value as? String else { return }
            self?.afterLoad {
                self?.evaluate("window.__panel.setSource(\(JSONBridge.quoted(html)))")
            }
        }
    }

    func clearNetwork() {
        afterLoad { [weak self] in self?.evaluate("window.__panel.clearNetwork()") }
    }

    func appendNetwork(_ entry: NetworkEntry, pageURL: String) {
        let payload: [String: Any] = [
            "method": entry.method,
            "url": entry.url,
            "displayUrl": NetexURL.shortened(entry.url, pageURL: pageURL),
            "status": entry.status,
            "body": entry.body,
            "contentType": entry.contentType,
            "durationMs": entry.durationMs,
            "size": entry.size,
            "requestId": entry.requestID,
            "pending": entry.pending
        ]
        afterLoad { [weak self] in
            self?.evaluate("window.__panel.appendNetworkRow(\(JSONBridge.jsonString(payload)))")
        }
    }

    func appendNetwork(_ entries: [NetworkEntry], pageURL: String) {
        entries.forEach { appendNetwork($0, pageURL: pageURL) }
    }

    func appendConsole(_ entry: ConsoleEntry) {
        let payload: [String: Any] = [
            "level": entry.level,
            "segments": entry.segments,
            "sourceId": entry.source,
            "line": entry.line
        ]
        afterLoad { [weak self] in
            self?.evaluate("window.__panel.appendConsoleEntry(\(JSONBridge.jsonString(payload)))")
        }
    }

    func appendConsole(_ entries: [ConsoleEntry]) {
        entries.forEach { appendConsole($0) }
    }

    func deliverEvalResult(id: String, resultJSON: String) {
        afterLoad { [weak self] in
            self?.evaluate("window.__panel.deliverEvalResult(\(JSONBridge.quoted(id)), \(JSONBridge.quoted(resultJSON)))")
        }
    }

    func deliverBlob(rid: String, dataURL: String) {
        afterLoad { [weak self] in
            self?.evaluate("window.__panel.deliverBlob(\(JSONBridge.quoted(rid)), \(JSONBridge.quoted(dataURL)))")
        }
    }

    private func applyTheme() {
        let dark = UIScreen.main.traitCollection.userInterfaceStyle == .dark
        let fg = dark ? "#cccccc" : "#222222"
        let panelBg = dark ? "#1f1f23" : "#f6f6f7"
        let viewerBg = dark ? "#2c2c30" : "#ffffff"
        evaluate("window.__panel.setTheme({ dark: \(dark), fg: '\(fg)', panelBg: '\(panelBg)', viewerBg: '\(viewerBg)' })")
    }

    private func afterLoad(_ action: @escaping () -> Void) {
        if shellLoaded { action() } else { pending.append(action) }
    }

    private func evaluate(_ script: String) {
        panelView?.evaluateJavaScript(script)
    }
}
